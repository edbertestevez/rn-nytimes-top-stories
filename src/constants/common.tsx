export interface IKeyLabel {
  key: string;
  label: string;
}

export enum RequestMethods {
  GET = 'get',
  POST = 'post',
  PATCH = 'patch',
  DELETE = 'delete',
}
