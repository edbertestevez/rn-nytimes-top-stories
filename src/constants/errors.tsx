export const ERROR_NO_NETWORK_CONNECTION = 'No Internet Connection';
