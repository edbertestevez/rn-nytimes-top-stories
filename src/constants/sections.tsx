const sections = [
  {key: 'arts', label: 'arts'},
  {key: 'automobiles', label: 'automobiles'},
  {key: 'books', label: 'books'},
  {key: 'business', label: 'business'},
  {key: 'fashion', label: 'fashion'},
  {key: 'food', label: 'food'},
  {key: 'health', label: 'health'},
  {key: 'home', label: 'home'},
  {key: 'insider', label: 'insider'},
  {key: 'magazine', label: 'magazine'},
  {key: 'movies', label: 'movies'},
  {key: 'nyregion', label: 'nyregion'},
  {key: 'obituaries', label: 'obituaries'},
  {key: 'opinion', label: 'opinion'},
  {key: 'politics', label: 'politics'},
  {key: 'realestate', label: 'realestate'},
  {key: 'science', label: 'science'},
  {key: 'sports', label: 'sports'},
  {key: 'sundayreview', label: 'sundayreview'},
  {key: 'technology', label: 'technology'},
  {key: 'theater', label: 'theater'},
  {key: 't-magazine', label: 't-magazine'},
  {key: 'travel', label: 'travel'},
  {key: 'upshot', label: 'upshot'},
  {key: 'us', label: 'us'},
  {key: 'world', label: 'world'},
];

export default sections;
