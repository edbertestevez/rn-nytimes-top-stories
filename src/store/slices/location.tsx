import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {ISelectItem} from '../../types/Input';

interface ILocationState {
  countries: Array<ISelectItem>;
}

const defaultCountry = [
  {
    label: 'All Location',
    value: '',
  },
];

const initialState: ILocationState = {
  countries: defaultCountry,
};

interface IPayload {
  countries: Array<ISelectItem>;
}

export const locationSlice = createSlice({
  name: 'location',
  initialState,
  reducers: {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    reset: (state) => {
      state = initialState;
    },
    setCountries: (state: ILocationState, action: PayloadAction<IPayload>) => {
      state.countries = [...defaultCountry, ...action.payload.countries];
    },
  },
});

export const locationSliceActions = {
  ...locationSlice.actions,
};
