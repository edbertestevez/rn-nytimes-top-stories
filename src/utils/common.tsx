import {Platform} from 'react-native';

export const isNullEmptyOrUndefined = (value: any) => {
  return (
    typeof value === 'undefined' ||
    value === null ||
    value === '' ||
    (typeof value === 'object' && Object.keys(value).length <= 0)
  );
};

export const isPlatformAndroid = () => {
  return Platform.OS === 'android';
};

export const isPlatformIOS = () => {
  return Platform.OS === 'ios';
};
