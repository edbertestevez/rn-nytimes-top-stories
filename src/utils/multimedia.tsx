import {IMultimedia} from '../types/News';
import {isNullEmptyOrUndefined} from './common';

export enum MediaSize {
  SUPER_JUMBO = 'superJumbo',
  STANDARD = 'Standard Thumbnail',
  LARGE = 'thumbLarge',
  MEDIUM = 'mediumThreeByTwo210',
  NORMAL = 'Normal',
}

export const get_multimedia_by_format = (
  media: Array<IMultimedia>,
  size: MediaSize,
) => {
  if (isNullEmptyOrUndefined(media)) {
    return null;
  }

  return media.filter((item: IMultimedia) => item.format === size)[0];
};
