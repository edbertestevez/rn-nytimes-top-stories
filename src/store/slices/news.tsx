import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {sectionKeys} from '../../constants/sections';
import {News} from '../../types/News';

interface INews {
  list: Array<News>;
  sectionFilter: string;
  keywordFilter: string;
}

const initialState: INews = {
  list: [],
  sectionFilter: sectionKeys.world,
  keywordFilter: '',
};

export const newsSlice = createSlice({
  name: 'news',
  initialState,
  reducers: {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    reset: (state) => {
      state = initialState;
    },
    setList: (state, action: PayloadAction<Array<News>>) => {
      state.list = action.payload;
    },
    setSectionFilter: (state, action: PayloadAction<string>) => {
      state.sectionFilter = action.payload;
    },
    setKeywordFilter: (state, action: PayloadAction<string>) => {
      state.keywordFilter = action.payload;
    },
  },
});

export const newsSliceActions = {
  ...newsSlice.actions,
};
