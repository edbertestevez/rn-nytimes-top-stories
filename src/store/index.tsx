import AsyncStorage from '@react-native-async-storage/async-storage';
import {
  configureStore,
  getDefaultMiddleware,
  combineReducers,
} from '@reduxjs/toolkit';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';

//Slices
import {newsSlice} from './slices/news';

const reducers = combineReducers({
  news: newsSlice.reducer,
});

const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
};

const persistedReducer = persistReducer(persistConfig, reducers);

const store = configureStore({
  reducer: persistedReducer,
  middleware: getDefaultMiddleware({
    immutableCheck: false,
    serializableCheck: {
      ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
    },
  }),
});

const persistor = persistStore(store);

export {store, persistor};

export type RootState = ReturnType<typeof store.getState>;
