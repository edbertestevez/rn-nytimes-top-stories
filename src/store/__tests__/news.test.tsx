import configureMockStore from 'redux-mock-store';
import {initialState, newsSliceActions, INews} from '../slices/news';
import thunk from 'redux-thunk';
import {AppDispatch} from '../../store/index';
import {sectionKeys} from '../../constants/sections';
import {MediaSize} from '../../utils/multimedia';

const middlewares: any = [thunk];
const mockStore = configureMockStore(middlewares);

describe('[Redux] News slice testing', () => {
  let store: any;
  let dispatch: AppDispatch;

  beforeEach(() => {
    store = mockStore(initialState);
    store.clearActions();
    dispatch = store.dispatch;
  });

  let mockListFilter = {
    list: [
      {
        title: 'Test',
        abstract: 'Test',
        url: 'Test',
        uri: 'Test',
        byline: 'Test',
        published_date: new Date('2020-12-16T06:46:53.375Z'),
        multimedia: [
          {
            url:
              'https://static01.nyt.com/images/2020/12/15/arts/best-comedy-2020-topart/best-comedy-2020-topart-superJumbo.jpg',
            format: MediaSize.SUPER_JUMBO,
            type: 'image',
            caption: 'Caption',
          },
          {
            url:
              'https://static01.nyt.com/images/2020/12/15/arts/best-comedy-2020-topart/best-comedy-2020-topart-thumbStandard.jpg',
            format: MediaSize.STANDARD,
            type: 'image',
            caption: 'Caption',
          },
          {
            url:
              'https://static01.nyt.com/images/2020/12/15/arts/best-comedy-2020-topart/best-comedy-2020-topart-thumbLarge.jpg',
            format: MediaSize.LARGE,
            type: 'image',
            caption: 'Caption',
          },
          {
            url:
              'https://static01.nyt.com/images/2020/12/15/arts/best-comedy-2020-topart/best-comedy-2020-topart-mediumThreeByTwo210.jpg',
            format: MediaSize.MEDIUM,
            type: 'image',
            caption: 'Caption',
          },
        ],
      },
    ],
  };

  it('check intial news state', () => {
    let news: INews = store.getState();

    expect(news.sectionList).toEqual({});
    expect(news.articleSearchList).toEqual([]);
    expect(news.sectionFilter).toEqual(sectionKeys.world);
    expect(news.keywordFilter).toEqual('');
    expect(news.locationFilter).toEqual('');
  });

  it('creates <news>/reset when dispatch is successful', () => {
    dispatch(newsSliceActions.reset());

    expect(store.getActions()).toEqual([
      {
        payload: undefined,
        type: 'news/reset',
      },
    ]);

    expect(store.getActions()).toMatchSnapshot();

    expect(store.getState()).toEqual(initialState);
  });

  it('creates <news>/setSectionFilter when dispatch is successful', () => {
    dispatch(newsSliceActions.setSectionFilter('Test'));

    expect(store.getActions()).toEqual([
      {
        payload: 'Test',
        type: 'news/setSectionFilter',
      },
    ]);

    expect(store.getActions()).toMatchSnapshot();
  });

  it('creates <news>/setKeywordFilter when dispatch is successful', () => {
    dispatch(newsSliceActions.setKeywordFilter('Test'));

    expect(store.getActions()).toEqual([
      {
        payload: 'Test',
        type: 'news/setKeywordFilter',
      },
    ]);

    expect(store.getActions()).toMatchSnapshot();
  });

  it('creates <news>/setLocationFilter when dispatch is successful', () => {
    dispatch(newsSliceActions.setLocationFilter('Test'));

    expect(store.getActions()).toEqual([
      {
        payload: 'Test',
        type: 'news/setLocationFilter',
      },
    ]);

    expect(store.getActions()).toMatchSnapshot();
  });

  it('creates <news>/setList when dispatch is successful', () => {
    dispatch(newsSliceActions.setList(mockListFilter));

    expect(store.getActions()).toEqual([
      {
        payload: mockListFilter,
        type: 'news/setList',
      },
    ]);

    expect(store.getActions()).toMatchSnapshot();
  });

  it('creates <news>/setArticleSearchList when dispatch is successful', () => {
    dispatch(newsSliceActions.setArticleSearchList(mockListFilter));

    expect(store.getActions()).toEqual([
      {
        payload: mockListFilter,
        type: 'news/setArticleSearchList',
      },
    ]);

    expect(store.getActions()).toMatchSnapshot();
  });
});
