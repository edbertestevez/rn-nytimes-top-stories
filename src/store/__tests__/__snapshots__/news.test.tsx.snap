// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Redux] News slice testing creates <news>/reset when dispatch is successful 1`] = `
Array [
  Object {
    "payload": undefined,
    "type": "news/reset",
  },
]
`;

exports[`[Redux] News slice testing creates <news>/setArticleSearchList when dispatch is successful 1`] = `
Array [
  Object {
    "payload": Object {
      "list": Array [
        Object {
          "abstract": "Test",
          "byline": "Test",
          "multimedia": Array [
            Object {
              "caption": "Caption",
              "format": "superJumbo",
              "type": "image",
              "url": "https://static01.nyt.com/images/2020/12/15/arts/best-comedy-2020-topart/best-comedy-2020-topart-superJumbo.jpg",
            },
            Object {
              "caption": "Caption",
              "format": "Standard Thumbnail",
              "type": "image",
              "url": "https://static01.nyt.com/images/2020/12/15/arts/best-comedy-2020-topart/best-comedy-2020-topart-thumbStandard.jpg",
            },
            Object {
              "caption": "Caption",
              "format": "thumbLarge",
              "type": "image",
              "url": "https://static01.nyt.com/images/2020/12/15/arts/best-comedy-2020-topart/best-comedy-2020-topart-thumbLarge.jpg",
            },
            Object {
              "caption": "Caption",
              "format": "mediumThreeByTwo210",
              "type": "image",
              "url": "https://static01.nyt.com/images/2020/12/15/arts/best-comedy-2020-topart/best-comedy-2020-topart-mediumThreeByTwo210.jpg",
            },
          ],
          "published_date": 2020-12-16T06:46:53.375Z,
          "title": "Test",
          "uri": "Test",
          "url": "Test",
        },
      ],
    },
    "type": "news/setArticleSearchList",
  },
]
`;

exports[`[Redux] News slice testing creates <news>/setKeywordFilter when dispatch is successful 1`] = `
Array [
  Object {
    "payload": "Test",
    "type": "news/setKeywordFilter",
  },
]
`;

exports[`[Redux] News slice testing creates <news>/setList when dispatch is successful 1`] = `
Array [
  Object {
    "payload": Object {
      "list": Array [
        Object {
          "abstract": "Test",
          "byline": "Test",
          "multimedia": Array [
            Object {
              "caption": "Caption",
              "format": "superJumbo",
              "type": "image",
              "url": "https://static01.nyt.com/images/2020/12/15/arts/best-comedy-2020-topart/best-comedy-2020-topart-superJumbo.jpg",
            },
            Object {
              "caption": "Caption",
              "format": "Standard Thumbnail",
              "type": "image",
              "url": "https://static01.nyt.com/images/2020/12/15/arts/best-comedy-2020-topart/best-comedy-2020-topart-thumbStandard.jpg",
            },
            Object {
              "caption": "Caption",
              "format": "thumbLarge",
              "type": "image",
              "url": "https://static01.nyt.com/images/2020/12/15/arts/best-comedy-2020-topart/best-comedy-2020-topart-thumbLarge.jpg",
            },
            Object {
              "caption": "Caption",
              "format": "mediumThreeByTwo210",
              "type": "image",
              "url": "https://static01.nyt.com/images/2020/12/15/arts/best-comedy-2020-topart/best-comedy-2020-topart-mediumThreeByTwo210.jpg",
            },
          ],
          "published_date": 2020-12-16T06:46:53.375Z,
          "title": "Test",
          "uri": "Test",
          "url": "Test",
        },
      ],
    },
    "type": "news/setList",
  },
]
`;

exports[`[Redux] News slice testing creates <news>/setLocationFilter when dispatch is successful 1`] = `
Array [
  Object {
    "payload": "Test",
    "type": "news/setLocationFilter",
  },
]
`;

exports[`[Redux] News slice testing creates <news>/setSectionFilter when dispatch is successful 1`] = `
Array [
  Object {
    "payload": "Test",
    "type": "news/setSectionFilter",
  },
]
`;
